File: app/src/main/java/com/sdex/activityrunner/util/RunActivityTask.java
package com.sdex.activityrunner.util;

import android.content.ComponentName;
import android.os.AsyncTask;
import android.util.Log;

import com.topjohnwu.superuser.Shell;

public class RunActivityTask extends AsyncTask<Void, Void, Integer> {

  private static final String TAG = "RunActivityTask";
  private final ComponentName componentName;

  public RunActivityTask(ComponentName componentName) {
    this.componentName = componentName;
  }

  @Override
  protected Integer doInBackground(Void[] params) {
    String className = componentName.getClassName();
    if (className.contains("$")) {
      className = className.replace("$", "\\$");
    }

    Log.d(TAG, "class name: " + className);
    try {
      final String command = "am start -n " + componentName.getPackageName() + "/" + className;
      Shell.Sync.su(command);
    } catch (Exception e) {
      e.printStackTrace();
    }
    return 0;
  }
}

File: app/src/main/java/com/sdex/activityrunner/util/RunActivityTask.kt
package com.sdex.activityrunner.util

import android.content.ComponentName
import android.os.AsyncTask
import com.sdex.commons.analytics.AM_START_ACTIVITY
import com.sdex.commons.analytics.AnalyticsManager

class RunActivityTask(private val componentName: ComponentName) : AsyncTask<Void, Void, Int>() {

  override fun doInBackground(params: Array<Void>): Int? {
    var className = componentName.className
    if (className.contains("$")) {
      className = className.replace("$", "\\$")
    }

    try {
      val command = "am start -n " + componentName.packageName + "/" + className
      RootUtils.execute(command)
    } catch (e: Exception) {
      AnalyticsManager.logError(AM_START_ACTIVITY, componentName.packageName, e.message)
      e.printStackTrace()
    }

    return 0
  }
}
